ポイントボーナスキャンペーンの条件変更

naveqo opened the pull request:
## Pivotal
https://www.pivotaltracker.com/story/show/151940683

## 関連するPR/Issue

### 親PR
https://github.com/SunriseDigital/nukinavi_unified/pull/629

### 関連PR
https://github.com/SunriseDigital/cs-fuzoku/pull/235

## Page
https://shutoken.nukinavi.com/admin/master/point_campaign/

naveqo committed:
Make PR

naveqo committed:
テストコードとプロパティ追加

naveqo committed:
メソッドがNULLだった場合も通す

naveqo committed:
PropertyInfoキャスト用のメソッドを追加

naveqo committed:
テスト成功

naveqo committed:
例外の文言を変更

naveqo committed:
テストメソッド名の変更

naveqo committed:
assert追加、例外コメントにメソッド名かプロパティ名がくるように変更

gomo commented on Sdx/Scaffold/Config/Value.cs:
```diff
@@ -65,12 +65,25 @@ internal MethodInfo ToMethodInfo(Type type)
       {
         var method = type.GetMethod(value.ToString());
-        if (method == null)
+        return method;
```
ここは一度変数に入れる必要はなさそうですね。
代入にも計算量が1かかるので必要なのであればしない方がいいです。まあ、微々たるもんなんですけど。

まあ直すほどでもないけどついでがあったら直してもらおうかな。

```rb
return type.GetMethod(value.ToString());
```

🐒

naveqo replied:
確かに必要ないですね。。詰めが甘かったです。。。
修正承知しました。

gomo commented on Sdx/Scaffold/Config/Value.cs:
```diff
@@ -65,12 +65,25 @@ internal MethodInfo ToMethodInfo(Type type)
+      else
+      {
+        var property = type.GetProperty(value.ToString());
+        return property;
```
こちらもですね。まあ、上のコード使いまわしたからだけど。ただ、こういうことに神経を研ぎ澄ませとくのはいいことです。

gomo commented on Sdx/Scaffold/Manager.cs:
```diff
@@ -367,13 +367,30 @@ public void Save(Sdx.Db.Record record, NameValueCollection values, Sdx.Db.Connec
+              {
+                var propertyInfo = config["setter"].ToPropertyInfo(record.GetType());
+
+                args.Add(values);
```
`args.Add(values)`は必要なさそうですね。

上のメソッドでこれが必要な理由は、サブミットされた他の値をつかって何かをしたい時のためにメソッドに２つの引数を宣言して扱える様になってるからですね。

```cs
.Set("setter", new Sdx.Scaffold.Config.Value("FooBar"))
...
namespace Test.Orm
{
  public class Area : Sdx.Db.Record
  {
    public void FooBar(string value, NameValueCollection values)
    {
       //こういう風に宣言しておくと、全てのサブミットされた値が取れるんです。
    }
```

プロパティは値の代入しかないので
```cs
propertyInfo.SetValue(record, args[0]);
```

だけすればOKです。

😇

naveqo replied:
> 上のメソッドでこれが必要な理由は、サブミットされた他の値をつかって何かをしたい時のためにメソッドに２つの引数を宣言して扱える様になってるからですね。

何故`Count() == 2`なんだろう？って思ってました。
`methodInfo.GetParameters()`をダンプして謎が解けました。

> //こういう風に宣言しておくと、全てのサブミットされた値が取れるんです。

これが微妙に理解出来てないので後で聞きに行きます！！

gomo commented on UnitTest/Sdx/Scaffold/Manager.cs:
```diff
@@ -1492,5 +1492,65 @@ private void RunPostSaveHook(TestDb db)
+      post.Set("type", "new");
+      post.Set("code", "test");
+
+      string savedId;
```
この宣言は必要ないですね。C#の変数はブロック変数なのでusingの中で使うなら必要ないです。

下記の場合は必要です。

```cs

      string savedId;
      using (var conn = scaffold.Db.CreateConnection())
      {
        ...
        savedId = record.GetString("id");
      }

      var savedRecord = (new Test.Orm.Table.Area()).FetchRecordByPkey(conn, savedId);
      Assert.Equal("new", savedRecord.GetString("name"));
      Assert.Equal("test", savedRecord.GetString("code"));
```

コピー元もこうなってるので、コピー元もおそらくどっかのテストをコピーしたのでそうなってるんでしょうw
まあ、ついでがあったら直しておいてください。

知識としては知っておいてください。

ちなみに、変数のスコープは言語によって微妙に違うので気をつけてくださいね。C#は、という話です。PHPやJSなど型にゆるい言語はifとかforなどのブロックの中で宣言した変数が外で使えちゃったりします。関数ブロックはつかません。そういう言語はC#のブロックスコープと対比して関数スコープと言ったりします。

🐒

naveqo replied:
`savedId`変数を使わない形で修正しました。

gomo commented on UnitTest/Sdx/Scaffold/Manager.cs:
```diff
@@ -1492,5 +1492,65 @@ private void RunPostSaveHook(TestDb db)
+
+      var query = new NameValueCollection();
+      var post = new NameValueCollection();
+      post.Set("type", "new");
```
とりあえず、ここまでできたので、ここを配列に変更してみてください。

```cs
post.Add("type", "foo")
post.Add("type", "bar")
post.Add("type", "baz")
```

でアサート（保存後の結果）は

```
Assert.Equal("new", savedRecord.GetString("foo_bar_baz"));
```
となる様にテストを変更してみてください。

😇

naveqo committed:
不要な代入をカット

naveqo committed:
プロパティは値の代入しかないのでListへの追加は不要

naveqo committed:
ダミーデータを配列に変更　余分な変数の宣言をカット

naveqo committed:
プロパティに配列をセットして文字列に変更してからカラムにしまう

gomo commented on Sdx/Scaffold/Manager.cs:
```diff
@@ -382,10 +382,7 @@ public void Save(Sdx.Db.Record record, NameValueCollection values, Sdx.Db.Connec
-
-                propertyInfo.SetValue(record, args[0]);
-
+                propertyInfo.SetValue(record, values.GetValues(columnName));
```
おしいなあw

これをしちゃうと、前の文字列が動かなくなっちゃうんですよね。そういう意味ではテストをもう一個追加するとよかったですね。そうすれば動かなくなったことに気づいたので。

ヒントはここです。
https://github.com/SunriseDigital/cs-sdx/blob/65ebcea36887df5da70e4936c3ae1e7406194444/Sdx/Scaffold/Manager.cs#L361-L368

配列用と文字列用両方テストしましょうか？前のテストもまるっとコピーして持って来て、ヒントを参考にテストが両方通る様にしてください。

https://github.com/SunriseDigital/cs-sdx/blob/e83f2a678d3c2f911f628b477609abf9ab6f0db2/UnitTest/Sdx/Scaffold/Manager.cs#L1496-L1552

😇

naveqo replied:
ああ。。なるほど。確かにですね。
承知しました！

naveqo committed:
Merge branch 'master' into proj/addConditionNewCustomersCoupon

naveqo committed:
文字列と配列を分けてテストするためテストコードを追加

naveqo committed:
渡す値を配列から文字列に

naveqo committed:
文字列を受けてる用のプロパティ追加

naveqo committed:
プロパティの型を見てSetValueする値を変える

gomo commented on Sdx/Scaffold/Manager.cs:
```diff
@@ -382,7 +382,14 @@ public void Save(Sdx.Db.Record record, NameValueCollection values, Sdx.Db.Connec
                 var propertyInfo = config["setter"].ToPropertyInfo(record.GetType());
 
-                propertyInfo.SetValue(record, values.GetValues(columnName));
+                if (propertyInfo.PropertyType.Name.Contains("[]"))
```
ああ、なるほど。こういう判断方法もないことはないけど・・・

ヒントで送ったコードを見て見てください。

https://github.com/SunriseDigital/cs-sdx/blob/dcf3b976b4f98ccdc9069723e1e690b712bc9b58/Sdx/Scaffold/Manager.cs#L361-L368

関数の時と判断基準が変わってしまうとややこしくないですか？関数の時も同じ様に変更するならなくはないですけど、わざわざ仕様を変更する積極的な理由がないですね。既存のコードにも影響するし。

あと、この送ったヒントは「関数の時は」というIF文に入ってないですよね。つまり、プロパティの時も使えるということです。で、その判断基準はどうなってるかというと、設定項目の中に`multiple`という値が`true`である場合は強制的に配列でセットするようになっています。スカッフォールドの使用時に明示的に指定するような仕様になっているということです。

プロパティの時も同じ方法が使えます。

```c#
      scaffold.FormList
        .Add(Sdx.Scaffold.Config.Item.Create()
          .Set("label", new Sdx.Scaffold.Config.Value("タイプ"))
          .Set("name", new Sdx.Scaffold.Config.Value("type"))
          .Set("multiple", new Sdx.Scaffold.Config.Value(true))
          .Set("setter", new Sdx.Scaffold.Config.Value("Types"))
        )
```

😇

naveqo replied:
まったくもって勘違いしておりました。。
承知しました！元に戻します！！

naveqo committed:
プロパティとメソッドの時とで判断基準が変わってしまうので元に戻した

naveqo committed:
Travisでエラーが出たので保存するカラムの値を変更

gomo reviewed (approved):
👍

naveqo committed:
セットされた値がNULLだった時の条件を追加

現状だとvalues[columnName]がNULLだった時にsetterを扱うブロック内に入らな
い為、NULLかつallowNullがTrueだった時だけNULLをセットするようにした。

naveqo committed:
NULLの条件をsetterを処理するブロック内で行うように変更

naveqo committed:
setterがセットされてない場合のNULL条件を追加

gomo commented on Sdx/Scaffold/Manager.cs:
```diff
@@ -353,44 +353,57 @@ public void Save(Sdx.Db.Record record, NameValueCollection values, Sdx.Db.Connec
+            {
+              var propertyInfo = config["setter"].ToPropertyInfo(record.GetType());
 
+              if (args != null) { 
```
こちらも同じくいらないさそう。


```c#
              var propertyInfo = config["setter"].ToPropertyInfo(record.GetType());
              propertyInfo.SetValue(record, args[0]);
```



あと細かいですけどc#のIFはこんな感じで書くことになってます。言語によって推奨の書き方が違って、今のところ一応合わせてるので気をつけてください。まあ、これだけだったら修正しろとは言いませんけど。ついでに直してもらおうかな。

```c#
if (args != nulll)
{

}
```

😇

naveqo replied:
> あと細かいですけどc#のIFはこんな感じで書くことになってます。言語によって推奨の書き方が違って、今のところ一応合わせてるので気をつけてください。まあ、これだけだったら修正しろとは言いませんけど。ついでに直してもらおうかな。

承知しました。
以後気をつけて書くようにします！

gomo commented on Sdx/Scaffold/Manager.cs:
```diff
@@ -353,44 +353,57 @@ public void Save(Sdx.Db.Record record, NameValueCollection values, Sdx.Db.Connec
+                args.Add(values);
+              }
 
+              if (args != null)
```
このIF文は必ずTRUEになりますね。
https://github.com/SunriseDigital/cs-sdx/blob/f6cb333afb5a10c49d186181a4cd0e50dffc418e/Sdx/Scaffold/Manager.cs#L356
ここで初期化してるので。

たぶん
```c#
if (args[0] != null) { 

```
と勘違いかな？
でも、そもそもこのIF文はいらないんじゃないでしょうか？

```c#
            if (config["setter"].ToMethodInfo(record.GetType()) != null)
            {
              var methodInfo = config["setter"].ToMethodInfo(record.GetType());
             
              if (methodInfo.GetParameters().Count() == 2)
              {
                args.Add(values);
              }

              methodInfo.Invoke(record, args.ToArray());
            }
```

どうでしょう？


😇

naveqo replied:
> このIF文は必ずTRUEになりますね。

確認したところ必ずNULLになりました。
確認不足でした、すみません。。

gomo commented on Sdx/Scaffold/Manager.cs:
```diff
@@ -353,44 +353,57 @@ public void Save(Sdx.Db.Record record, NameValueCollection values, Sdx.Db.Connec
+          }
+          else
+          {
+            if (values[columnName] != null) {
```
このIF文もいらないんじゃないかな？
これがNULLになるということは、checkbox/radio/selectタグの時だけで、それが空の時はカラムをNULLにしちゃって良さそうですし・・・

カラムがNOT NULLで、空の時に空文字や0に更新したいシチュエーションがないとも限らないけど、そういうのはカラムの設定値から取って自動で調整するのがいいのかな。それがベストですね。

コメント書いておいてください。

```c#
// チェックボックスなどで`values[columnName]`がNULLの時、カラムをNULLに更新しようとします。
// TODO テーブルからカラムの設定を取り出してヌルを許容していない場合、型に応じて0か空文字などを入れる必要があります。
record.SetValue(columnName, values[columnName]);
```

😇

naveqo replied:
承知しました。

naveqo committed:
scaffold.Manager.Saveの処理変更したらパスワードがNULLの時にコケたのでテ
スト側の処理を変更

naveqo committed:
既存のテストコードがコケたので通るように修正

setterに渡す値をNULLではなく空文字で渡すようにしました

naveqo committed:
不要なIF文をカット

naveqo committed:
IsSecretを非推奨に

gomo reviewed (approved):
👍
